tests/test_component.py

```diff
@@ -1429,260 +1430,555 @@ class Root(Component):
+
+        rendered = SimpleComponent.render()
         assertHTMLEqual(
             rendered,
             """
-            Chocolate cookie recipe:
-            args: []
-            kwargs: {}
+            text
             ---
-            from_on_before:
+            FROM_ON_BEFORE
             ---
-            Hello from nested
-            <div data-djc-id-ca1bc3e data-djc-id-ca1bc40>
-                Hello from simple
-            </div>
+            FROM_ON_RENDER_PRE
             """,
         )
 
-    def test_on_render_before_after_same_context(self):
-        context_in_before = None
-        context_in_after = None
-
-        @register("nested")
-        class NestedComponent(Component):
+    def test_on_render_no_yield(self):
+        class SimpleComponent(Component):
             template: types.django_html = """
-                {% load component_tags %}
-                Hello from nested
-                <div>
-                    {% slot "content" default / %}
-                </div>
+                text
             """
 
+            def on_render(self, context: Context, template: Template):
+                return "OVERRIDDEN"
+
+        rendered = SimpleComponent.render()
+        assert rendered == "OVERRIDDEN"
+
+    def test_on_render_reraise_error(self):
+        registry.register("broken", self._gen_broken_component())
+
         class SimpleComponent(Component):
             template: types.django_html = """
-                {% load component_tags %}
-                args: {{ args|safe }}
-                kwargs: {{ kwargs|safe }}
-                ---
-                from_on_after: {{ from_on_after }}
-                ---
-                {% component "nested" %}
-                    Hello from simple
-                {% endcomponent %}
+                {% component "broken" / %}
             """
 
-            def get_template_data(self, args, kwargs, slots, context):
-                return {
-                    "args": args,
-                    "kwargs": kwargs,
-                }
+            def on_render(self, context: Context, template: Template):
+                html, error = yield template.render(context)
 
-            def on_render_before(self, context: Context, template: Template) -> None:
-                context["from_on_before"] = ":)"
-                nonlocal context_in_before
-                context_in_before = context
+                raise error from None  # Re-raise original error
 
-            # Check that modifying the context or template does nothing
-            def on_render_after(self, context: Context, template: Template, html: str) -> None:
-                context["from_on_after"] = ":)"
-                nonlocal context_in_after
-                context_in_after = context
+        with pytest.raises(ValueError, match=re.escape("BROKEN")):
+            SimpleComponent.render()
+
+    @djc_test(
+        parametrize=(
+            ["template", "action", "method"],
```

Comment:
And here I added tests for checking of `on_render_before()` / `on_render()` / `on_render_after()` behave when:
1. They receive error
2. They render template that contains error
3. They render valid template

vs

- When `on_render` / `on_render_after` should override the HTML
- When `on_render` / `on_render_after` should override the error
- Do nothing